spicy/toolchain/src/compiler/parser/parser.yy

```diff
@@ -383,6 +383,12 @@ type_decl     : opt_linkage TYPE scoped_id '='   { _docs.emplace_back(driver->do
                                                        u->setAttributes(builder->context(), $7);
                                                        $7 = {}; // don't associate with declaration
                                                        }
+                                                   // Type decls can have attributes, but only for certain types
```

Comment:
This is one way to do it and it gets rid of the hacky stringification which had to use in the parser before since we didn't have enough context. It was hoping for a way to have this encoded explicitly in the parsing rules, but I can see how that is a bigger and more tricky change due to tech debt in the parser and think this is still an improvement.

Nit: you matched the indention in this rule, but it seems it was already inconsistent with everything else we have in this file. WDYT about the following instead?

```cpp
                qtype opt_attributes ';'         { if ( auto u = $6->type()->tryAs<type::Unit>(); u && $7 && *$7 ) {
                                                      u->setAttributes(builder->context(), $7);
                                                      $7 = {}; // don't associate with declaration
                                                   }

                                                   // Type decls can have attributes, but only for certain types
                                                   if ( $7 && !$7->attributes().empty() ) {
                                                      auto ty = $6->type();
                                                      if ( ! (ty->isA<hilti::type::Struct>() || ty->isA<hilti::type::Enum>() || ty->isA<hilti::type::Bitfield>()) )
                                                          error(@7, "attributes are not allowed on type aliases");
                                                   }

                                                   $$ = builder->declarationType(std::move($3), std::move($6), std::move($7), std::move($1), __loc__);
                                                   $$->setDocumentation(_docs.back());
                                                   _docs.pop_back();
                                                 }
```

Reply:
yeah good call, honestly these Bison parsers go so far beyond the width of my editor that at all looks like a mess when programming anyway so it's tough to tell.

And don't worry, I was also hoping for that. I think depending on how #1901 gets implemented, it could become easier/better to make that change. I was going to delay until after that, but figured this solution is easy enough to put in before while it's on the top of my mind.

My other attempt at cleaning that up didn't exactly go smoothly :)